cilium/CFP-33716-packet-tracing.md

```diff
@@ -0,0 +1,636 @@
+
+Several edge cases are considered during parsing:
+- **No IP Options:** If the packet header indicates there are no options, parsing stops early, avoiding unnecessary processing.
+- **Invalid `ip_trace_id`:** invalid IDs are handled by setting the trace ID to 0.
+
+### Section 3: `ip_trace_id` state
+
+Creating a per-CPU array-map allows for storage of each packet’s `ip_trace_id` as it is processed. When the feature gate is disabled (see Section 2: Feature enablement), the map is not rendered, avoiding space inefficiencies.
+
+```c
+#ifdef ENABLE_PACKET_IP_TRACING 
+struct {
+    __uint(type, BPF_MAP_TYPE_PERCPU_ARRAY);
+    __uint(max_entries, 1);
+    __type(key, __u32); // only one key
+    __type(value, __u64); // ip_trace_id type
+    __uint(pinning, LIBBPF_PIN_BY_NAME);
+} ip_trace_id_map __section_maps_btf;
+...
+// remaining declarations 
+#endif
+```
+
+**Pros**
+- Non-competitive storage; packet tracing ID is stored outside of the skb struct which has limited room for custom attributes.
+
+**Considerations**
+- Must ensure that data is zero'ed correctly.
+- This is universal storage and not specifically stored in the SKB struct, so it must be zero'ed when a new packet is processed.
+- At an ingress point, if there is no trace ID stored in a packet, the stored `ip_trace_id` is zero'ed from the array-map.
+
+### Section 4: Event integration
+
+Include ID as event field for trace and drop structs and the flow message (for Hubble integration):
+
+```c
+struct trace_notify {
+    // Existing struct members
+    ...
+    __u64      ip_trace_id; // new trace ID field
+    // Remaining struct members
+    ...
+};
+```
+```c
+struct drop_notify {
+    // Existing struct members
+    ...
+    __u64      ip_trace_id; // new trace ID field
+    // Remaining struct members
+    ...
+};
+```
+
+```go
+type Flow struct {
+    // Existing struct members
+    ...
+    IpTraceID IpTraceID 
+    `protobuf:"varint,38,opt,name=ip_trace_id,json=ipTraceId,proto3" json:"ip_trace_id,omitempty"` // new trace ID field
+    // Remaining struct members
+    ...
+}
+
+type IpTraceID struct {
+    TraceID      uint64 `protobuf:"varint,1,opt,name=trace_id,json=traceId,proto3" json:"trace_id,omitempty"`
+    IpOptionType uint8  `protobuf:"varint,2,opt,name=ip_option_type,json=ipOptionType,proto3" json:"ip_option_type,omitempty"`
+}
+```
+
+#### Tagged-flow ring buffer
+
+For tagged flows, there is a separate ring buffer to ensure that the processing of these events does not interfere with other events generated by Cilium. This separation helps maintain the performance of the system and mitigate concerns for tagged flows overflowing the event ring buffer, especially in high-traffic environments.
```

Comment:
What exactly is the purpose of the extra ring buffer? Is it just sharing the trace IDs to userspace?

A few reasons to ask this:
- Does userspace later need to correlate events to understand what the trace_ids mean? Does that add extra complication?
- If the class of information this generates is similar to the class for trace messages, then we could integrate the info into the existing events and require fewer userspace wakeups / event parsing+handling iterations. On the other hand if it is clearly a different priority/class of info vs. existing trace messages, then separating it can make sense. We do this today with some signals like when the CT/NAT maps become full, as those signals are more important; having a dedicated channel for those can make them more reliable.

Reply:
The extra ring buffer is a consideration pointed out by @sypakine and @pchaigno to ensure that events for non-tagged traffic don't get overrun by events for tagged traffic.

To your points:
1. The userspace will not need to correlate events to understand the trace_ids. It should be indistinguishable to a user that some events were pushed to the tagged flow ring buffer as opposed to the normal event ring buffer. This is to ensure that the current event generation is not overrun by tagged flows when this feature is enabled. 
2. As of the implementation I have right now, I do just that integration of these trace_ids into existing structs (https://github.com/cilium/cilium/pull/33957). These are implementation details that @sypakine will be taking over this work as I leave my internship.

Reply:
Hmm OK, I guess it would probably help to have a writeup of the exact patterns expected for usage of each ringbuffer and how the trace_ids flow through, depending on the traffic patterns.

For instance, let's say a packet with a trace_id arrives. Is only one trace_to event generated? And is that now generated to the new ringbuffer and not the old one? Or both?

Separately there's also the monitor_aggregation setting which avoids sending too many messages to userspace, so it'd be worth considering how that option intersects with this new option. I guess that's related to the thread with Paul about overloading; tagged packets could end up increasing userspace CPU usage for instance.

Reply:
I don't know if my feedback above is necessarily immediately actionable, and may be partially up to implementation details/discussion, but I would be happy to merge this maybe just with an extra line to highlight that this part is not fully resolved (unless we think we're likely to drive a clear conclusion from this discussion):

```suggestion
For tagged flows, there is a separate ring buffer to ensure that the processing of these events does not interfere with other events generated by Cilium. This separation helps maintain the performance of the system and mitigate concerns for tagged flows overflowing the event ring buffer, especially in high-traffic environments.

UNRESOLVED: How does the agent choose when to use this new ringbuffer or not? How does this integrate with Monitor Aggregation?
```

Reply:
Fixed and pushed this change.